ansible/roles/atmo-ssh-setup/tasks/main.yml

```diff
@@ -1,145 +1,73 @@
-  with_items: '{{ SSHKEYS }}'
-  when: default_remote_user|success and SSHKEYS is defined
-
-- name: remove ssh keys from authorized_keys
-  authorized_key: user=root key="{{ item }}" state=absent
-  with_items: '{{ SSH_KEYS_TO_REMOVE }}'
-  when: default_remote_user|success and SSH_KEYS_TO_REMOVE is defined
-
-# at this point, we could disable cloud user, but for now, let's leave in.
-
-###
-# OTHER SSH CONFIG CHANGES
-##
-- name: Remove all AllowUsers lines to /etc/sshd_config
-  lineinfile: dest=/etc/ssh/sshd_config regexp="{{ item }}" state=absent
-  with_items:
-    - AllowUsers
-    - AllowGroups
-  tags:
-    - debug
-    - delete
-    
-- name: /etc/ssh/sshd_config changes
-  lineinfile: dest=/etc/ssh/sshd_config backup=yes regexp="{{ item.regexp }}" line="{{ item.line }}"
-  with_items:
-    - { regexp: '^PermitRootLogin', line: 'PermitRootLogin without-password'}
-    - { regexp: '^UseDNS', line: 'UseDNS no'}
-    - { regexp: '^PasswordAuthentication', line: 'PasswordAuthentication yes'}
-
-- name: Append AllowGroups Line to /etc/ssh/sshd_config 
-  lineinfile: dest=/etc/ssh/sshd_config line="AllowGroups {{ SSH_ALLOW_GROUPS }}"
-  tags:
-    - debug
-
-- name: manual OS detection
-  stat: path=/etc/redhat-release
-  register: os_ver
-  tags:
-    - debug
-    - ssh-restart
-
-- name: restart ssh if ubuntu
-  service: name=ssh state=restarted
-  when: not os_ver.stat.exists
-  tags:
-    - debug
-    - ssh-restart
-
-- name: restart sshd if centos
-  service: name=sshd state=restarted
-  when: os_ver.stat.exists
-  tags:
-    - debug
-    - ssh-restart
+  tags: set_facts
+
+# SSH as different users to determine which can connect (and determine linux distro)
+- block:
+  - name: Test default ansible connection and register root_connection
+    local_action: >
+      command ssh {{ SSH_OPTIONS }} -p {{ ansible_port }} root@{{ vm_ip }} exit
+    register: root_connection
+
+  - name: Test CentOS connection and register centos_connection
+    local_action: >
+      command ssh {{ SSH_OPTIONS }} -p {{ ansible_port }} centos@{{ vm_ip }} exit
+    register: centos_connection
+    when: root_connection.rc != 0
+
+  - name: Test Ubuntu connection and register ubuntu_connection
+    local_action: >
+      command ssh {{ SSH_OPTIONS }} -p {{ ansible_port }} ubuntu@{{ vm_ip }} exit
+    register: ubuntu_connection
+    when: root_connection.rc != 0 and centos_connection.rc != 0
+  ignore_errors: true
+  tags: ssh
+
+# Set use_remote_user depending on which SSH task succeeded
+- block:
+  - name: Set use_remote_user variable to "centos"
+    set_fact: use_remote_user=centos
+    when: not centos_connection|skipped and centos_connection.rc == 0
+
+  - name: Set use_remote_user variable to "ubuntu"
+    set_fact: use_remote_user=ubuntu
+    when: not ubuntu_connection|skipped and ubuntu_connection.rc == 0
+
+  - name: Set use_remote_user variable to "root"
+    set_fact: use_remote_user=root
+    when: root_connection.rc == 0
+  tags: set_facts
```

Comment:
The rest of our playbooks depend on `use_remote_user` being defined. This block can leave the variable undefined. The last item in this block should include:
```
- fail:
    msg: "No remote connection established, atmo-ssh-setup can not continue."
  when: use_remote_user is not defined
```